domain/schema/model/sql/0020-network.sql

```diff
@@ -14,9 +14,304 @@ CREATE UNIQUE INDEX idx_cloud_service_application
 ON cloud_service (application_uuid);
 
 CREATE TABLE cloud_container (
+    -- one cloud container per net node.
     net_node_uuid TEXT NOT NULL PRIMARY KEY,
+    -- provider_id comes from the provider, no FK.
     provider_id TEXT NOT NULL,
     CONSTRAINT fk_cloud_container_net_node
     FOREIGN KEY (net_node_uuid)
     REFERENCES net_node (uuid)
 );
+
+CREATE UNIQUE INDEX idx_cloud_container_provider
+ON cloud_container (provider_id);
+
+CREATE TABLE cloud_container_port (
+    net_node_uuid TEXT NOT NULL,
+    port TEXT NOT NULL,
+    CONSTRAINT fk_cloud_container_port_net_node
+    FOREIGN KEY (net_node_uuid)
+    REFERENCES cloud_container (net_node_uuid),
+    PRIMARY KEY (net_node_uuid, port)
+);
+
+CREATE TABLE link_layer_device_type (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_link_layer_device_type_name
+ON link_layer_device_type (name);
+
+INSERT INTO link_layer_device_type VALUES
+(0, 'unknown'),
+(1, 'loopback'),
+(2, 'ethernet'),
+(3, '802.1q'),
+(4, 'bond'),
+(5, 'bridge'),
+(6, 'vxlan');
+
+CREATE TABLE virtual_port_type (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_virtual_port_type_name
+ON virtual_port_type (name);
+
+INSERT INTO virtual_port_type VALUES
+(0, 'nonvirtualport'),
+(1, 'openvswitch');
+
+CREATE TABLE link_layer_device (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    net_node_uuid TEXT NOT NULL,
+    name TEXT NOT NULL,
+    mtu INT,
+    mac_address TEXT NOT NULL,
+    -- provider_id comes from the provider, no FK.
+    provider_id TEXT,
+    -- one of loopback, ethernet, bridge etc.
+    device_type_id INT NOT NULL,
+    -- one of nonvirtualport, openvswitch
+    virtual_port_type_id INT NOT NULL,
+    -- true if the device should be activated on boot.
+    is_auto_start BOOLEAN NOT NULL DEFAULT false,
+    -- true when the device is up (enabled).
+    is_enabled BOOLEAN NOT NULL DEFAULT true,
+    -- the link layer device which is the parent to this device.
+    parent_device_uuid TEXT,
+    CONSTRAINT fk_link_layer_device_parent_device
+    FOREIGN KEY (parent_device_uuid)
+    REFERENCES link_layer_device (uuid),
+    CONSTRAINT fk_link_layer_device_net_node
+    FOREIGN KEY (net_node_uuid)
+    REFERENCES net_node (uuid),
+    CONSTRAINT fk_link_layer_device_device_type
+    FOREIGN KEY (device_type_id)
+    REFERENCES link_layer_device_type (id),
+    CONSTRAINT fk_link_layer_device_virtual_port_type
+    FOREIGN KEY (virtual_port_type_id)
+    REFERENCES virtual_port_type (id)
+);
+
+CREATE TABLE link_layer_device_ip_address (
```

Comment:
I think this bit is not quite right. The same IP address (the actual one designated by UUID, so a specific IP in a specific network) can't be assigned to more than one device.

Join tables are used for m-m, which this cannot be. Instead `ip_address` should just include `device_uuid`, which would also mean we can't orphan them.